Miscellaneous small fixes

tillarnold opened the pull request:
The changes to Prusti which I made in my recent PRs got a bit messy with the changes made in the upstream branch. So I am creating new PRs that should be easier to review and merge.

This PR is the first one and includes smaller changes that are needed for the later PRs:
- Enable the `log` feature of `tracing`. I did this because we are using `env_logger` and we do have configuration flags and code to use `env_logger` instead of the json based `tracing` output. However without the `log` feature of `tracing` this does not result in any output.
- Replace usage of `println` with `tracing::debug` calls
- Impure assert terminator
- The fix for phi destructing from #7 
- The temporary fix for allow tuple snapshots with different types to coexist from #8
- Some binops

zgrannan committed:
Introduce functionidentifier struct

zgrannan committed:
Fix bugs

zgrannan committed:
WIP

zgrannan committed:
WIP

zgrannan committed:
Cleanup

zgrannan committed:
refactor

zgrannan committed:
refactor

zgrannan committed:
Use type encoder in more places

zgrannan committed:
cleanup

zgrannan committed:
cleanup

JonasAlaif committed:
Merge branch 'rewrite-2023' into zgrannan/functionidentifier

JonasAlaif committed:
Resolve merge issues

JonasAlaif committed:
Clean up the type encoder

tillarnold committed:
Miscellaneous small fixes

zgrannan committed:
Update FunctionIdentifier

zgrannan committed:
Refactor to not use reify

zgrannan committed:
remove unused imports

JonasAlaif committed:
Add Predicate and Methods and clean things up

JonasAlaif committed:
Bugfix

JonasAlaif committed:
Merge branch 'zgrannan/functionidentifier' into till_rebuilt_1

JonasAlaif committed:
Merge fixes

JonasAlaif committed:
Builtin encoder improvements

JonasAlaif committed:
Merge branch 'rewrite-2023' into till_rebuilt_1

JonasAlaif committed:
Fix merge

JonasAlaif committed:
Bugfix

JonasAlaif committed:
Prevent use of `mk_func_app`

JonasAlaif committed:
Clean up `TerminatorKind::Assert`

Aurel300 commented on prusti-encoder/src/encoders/viper_tuple.rs:
```diff
@@ -150,7 +155,7 @@ impl TaskEncoder for ViperTupleEncoder {
                     name: domain_name,
                     typarams: vcx.alloc_slice(&typaram_names),
-                    axioms: vcx.alloc_slice(&[axiom]),
+                    axioms: if task_key == &0 {&[]} else {vcx.alloc_slice(&[axiom])},
```
Please extract this condition into a variable like `is_unit_tuple`. (Also spaces inside the braces are missing.)

JonasAlaif replied:
Agreed. @tillarnold this is not great code, if we have a unit slice then calculating and allocating everything for `axiom` is unnecessary - it's better to put this condition earlier on (see my latest push which fixes this)

Aurel300 commented on prusti-encoder/src/encoders/mir_builtin.rs:
```diff
@@ -15,12 +15,12 @@ pub enum MirBuiltinEncoderError {
-#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
 pub enum MirBuiltinEncoderTask<'tcx> {
     // TODO: which type? input or output? -> best to store both?
```
Stale TODO?

JonasAlaif replied:
Removed

Aurel300 commented on prusti-encoder/src/encoders/mir_builtin.rs:
```diff
@@ -15,12 +15,12 @@ pub enum MirBuiltinEncoderError {
-    BinOp(mir::BinOp, ty::Ty<'tcx>),
-    CheckedBinOp(mir::BinOp, ty::Ty<'tcx>),
+    UnOp(ty::Ty<'tcx>, mir::UnOp, ty::Ty<'tcx>),
+    // BinOp(mir::BinOp, ty::Ty<'tcx>),
```
Why are unchecked binops removed? Do we not need them at all?

JonasAlaif replied:
They weren't being used. However, now I've implemented all 3 properly according to spec.

JonasAlaif committed:
Fix review issue

JonasAlaif committed:
Various improvements

JonasAlaif commented:
Unfortunately this PR has ballooned a bit and now also includes `trusted` and external functions, a proper implementation of `mir_builtin` and bugfixes in `spec.rs`. @Aurel300 one more review?

JonasAlaif committed:
Fix join point algorithm

JonasAlaif committed:
Switch to `IndexVec`

JonasAlaif commented:
@Aurel300 please review the join point algorithm fix https://github.com/Aurel300/prusti-dev/pull/13/commits/8d6aca8016f8413d5185ebdaba7507c190d5a372 (essentially by removing it and doing it on the fly) and also the rest of the PR (i.e. the last 4 pushes: https://github.com/Aurel300/prusti-dev/compare/02ff0e2792272b0d3a33432a9d5611c2661f1bf8..524795898c4466ba4118bec6ef781295abc3695a)

Aurel300 commented on vir/src/data.rs:
```diff
@@ -83,7 +85,10 @@ pub enum ConstData {
 pub enum TypeData<'vir> {
-    Int,
+    Int {
+        bit_width: u8,
```
I find this strange: Viper integers don't have these attributes.

Aurel300 commented on prusti-encoder/src/encoders/mir_builtin.rs:
```diff
@@ -77,193 +76,277 @@ impl TaskEncoder for MirBuiltinEncoder {
+                            (Some(pre), val)
+                        }
+                        // Cannot overflow
+                        Div | Rem | BitXor | BitAnd | BitOr | Eq | Lt | Le | Ne | Ge | Gt | Offset =>
```
What about division by `-1`?

JonasAlaif replied:
Hmm good point. It seems that Rust guarantees that this never happens, even in release ([playground](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=d411cf1c5b21e857b6d01eb6bb17a579)). Not sure if we want to re-check this

JonasAlaif replied:
i.e. this is the same as division by `0` where we rely on Rust inserting an `assert` terminator before the division rather than rechecking things, or for overflows where Rust adds an `assert` after

JonasAlaif replied:
Though we probably should: both of [these](https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=58a9403964336dcb15ad2b4b5965d840) lead to UB (corresponding [llvm](https://llvm.org/docs/LangRef.html#sdiv-instruction) page)

Aurel300 commented on prusti-encoder/src/encoders/mir_pure.rs:
```diff
@@ -581,10 +588,15 @@ impl<'vir, 'enc> Encoder<'vir, 'enc>
             | mir::StatementKind::FakeRead(..)
-            | mir::StatementKind::AscribeUserType(..) => {}, // nop
+            | mir::StatementKind::AscribeUserType(..) 
+            | mir::StatementKind::PlaceMention(..)=> {}, // nop
```
```suggestion
            | mir::StatementKind::PlaceMention(..) => {}, // nop
```

Aurel300 commented on prusti-encoder/src/encoders/mir_pure.rs:
```diff
@@ -315,46 +315,51 @@ impl<'vir, 'enc> Encoder<'vir, 'enc>
-                    // rest is handled in a parent call.
-                    return stmt_update;
+            &mir::TerminatorKind::Goto { target } => {
+                match (dominators.immediate_dominator(target), branch_point) {
+                    // As soon as we are about to step to a bb where
+                    (Some(immediate_dominator), Some(branch_point))
+                        if immediate_dominator == branch_point =>
```
Can you not simply `branch_point == dominators.immediate_dominator(target)`?

JonasAlaif replied:
No, we do not want `true` if both are `None`

Aurel300 commented on prusti-encoder/src/encoders/mir_pure.rs:
```diff
@@ -315,46 +315,51 @@ impl<'vir, 'enc> Encoder<'vir, 'enc>
+                    _ => {
+                        // If we hit this todo, first verify that the join point
+                        // algorithm is working correctly. If it is then
+                        // consider continuing recursion here?
```
I'm not sure what you mean by "continuing recursion here".

Aurel300 commented on prusti-encoder/src/encoders/mir_pure.rs:
```diff
@@ -364,19 +369,18 @@ impl<'vir, 'enc> Encoder<'vir, 'enc>
-
-                // walk `start` -> `targets[i]` -> `join` for each target
+                // walk `curr` -> `targets[i]` -> `join` for each target. The
+                // join point is identified by reaching a bb where
+                // `dominators.immediate_dominator(bb) == curr`.
```
This describes *how* we find the join point, which I think belongs close to where that fact is relevant.

Aurel300 commented:
@JonasAlaif Mostly LGTM, see comments. For the new pure encoder algorithm, it's probably fine. I assume you tested with nested scenarios too (I'm not sure I understood that recursion comment)? Keeping track of the two endpoints of the subgraph seemed more intuitive, but probably not technically needed.

JonasAlaif committed:
Fix issues in `mir_builtin` and some lifetimes

- The `mir_builtin` should now be complete and not need to be changed
- Start work on separating `'tcx` and `'vir`

JonasAlaif committed:
Resolve review comments
